lib/dash/content_protection.js

```diff
@@ -252,16 +252,165 @@ shaka.dash.ContentProtection.parseFromRepresentation = function(
     // If we have filtered out all key-systems, throw an error.
     if (context.drmInfos.length == 0) {
       throw new shaka.util.Error(
-          shaka.util.Error.Severity.CRITICAL,
-          shaka.util.Error.Category.MANIFEST,
-          shaka.util.Error.Code.DASH_NO_COMMON_KEY_SYSTEM);
+        shaka.util.Error.Severity.CRITICAL,
+        shaka.util.Error.Category.MANIFEST,
+        shaka.util.Error.Code.DASH_NO_COMMON_KEY_SYSTEM);
     }
   }
 
   return repContext.defaultKeyId || context.defaultKeyId;
 };
 
 
+/**
+ * Gets a Widevine license URL from a content protection element
+ * containing a custom `ms:laurl` element
+ *
+ * @param {shaka.dash.ContentProtection.Element} element
+ * @return {string}
+ */
+shaka.dash.ContentProtection.getWidevineLicenseUrl = function(element) {
+  const mslaurlNode = shaka.util.XmlUtils.findChild(element.node, 'ms:laurl');
+  if (mslaurlNode) {
+    return mslaurlNode.getAttribute('licenseUrl') || '';
+  }
+  return '';
+};
+
+
+/**
+* @typedef {{
+*   type: number,
+*   length: number,
+*   value: string
+* }}
+*
+* @description
+* The parsed result of a PlayReady object record.
+*
+* @property {number} type
+*   Type of data stored in the record.
+* @property {number} length
+*   Size of the record in bytes.
+* @property {string} value
+*   Record content.
+*/
+shaka.dash.ContentProtection.PlayReadyRecord;
+
+/**
+ * Enum for PlayReady record types.
+ * @enum {number}
+ */
+shaka.dash.ContentProtection.PLAYREADY_RECORD_TYPES = {
+  RIGHTS_MANAGEMENT: 0x001,
+  RESERVED: 0x002,
+  EMBEDDED_LICENSE: 0x003,
+};
+
+
+/**
+* @param {Uint16Array} recordData
+* @param {number} recordCount
+* @return {!Array.<shaka.dash.ContentProtection.PlayReadyRecord>}
+* @private
+*/
+shaka.dash.ContentProtection.parseMsProRecords_ = function(
+  recordData, recordCount) {
+  let head = 0;
+  let records = [];
+  for (let i = 0; i < recordCount; i++) {
+    const recordRaw = recordData.subarray(head);
+    if (recordRaw.length > 1) {
+      const type = recordRaw[0];
+      const length = recordRaw[1];
+      const offset = 2;
+      const charCount = length / 2;
+      const end = charCount + offset;
+      // subarray end is exclusive
```

Comment:
Thanks for including this comment. It helps a lot.

Reply:
I can't take credit for that.  That's from the original PR.